pkg/controller/application/application_controller.go

```diff
@@ -18,45 +18,239 @@ package application
+
+		for _, u := range list.Items {
+			resource := u
+			resources = append(resources, &resource)
+		}
+	}
+	return resources
+}
+
+func (r *ApplicationReconciler) setOwnerRefForResources(ctx context.Context, ownerRef metav1.OwnerReference, resources []*unstructured.Unstructured) error {
+	for _, resource := range resources {
+		ownerRefs := resource.GetOwnerReferences()
+		ownerRefFound := false
+		for i, refs := range ownerRefs {
+			if ownerRef.Kind == refs.Kind &&
+				ownerRef.APIVersion == refs.APIVersion &&
+				ownerRef.Name == refs.Name {
+				ownerRefFound = true
+				if ownerRef.UID != refs.UID {
+					ownerRefs[i] = ownerRef
+				}
+			}
+		}
+
+		if !ownerRefFound {
+			ownerRefs = append(ownerRefs, ownerRef)
+		}
+		resource.SetOwnerReferences(ownerRefs)
+		err := r.Client.Update(ctx, resource)
+		if err != nil {
+			// We log this error, but we continue and try to set the ownerRefs on the other resources.
+			klog.Error(err, "ErrorSettingOwnerRef", "gvk", resource.GroupVersionKind().String(),
+				"namespace", resource.GetNamespace(), "name", resource.GetName())
+		}
+	}
+	return nil
+}
+
+func (r *ApplicationReconciler) objectStatuses(ctx context.Context, resources []*unstructured.Unstructured, errs *[]error) []appv1beta1.ObjectStatus {
+	var objectStatuses []appv1beta1.ObjectStatus
+	for _, resource := range resources {
+		os := appv1beta1.ObjectStatus{
+			Group: resource.GroupVersionKind().Group,
+			Kind:  resource.GetKind(),
+			Name:  resource.GetName(),
+			Link:  resource.GetSelfLink(),
+		}
+		s, err := status(resource)
+		if err != nil {
+			klog.Error(err, "unable to compute status for resource", "gvk", resource.GroupVersionKind().String(),
+				"namespace", resource.GetNamespace(), "name", resource.GetName())
+			*errs = append(*errs, err)
+		}
+		os.Status = s
+		objectStatuses = append(objectStatuses, os)
+	}
+	return objectStatuses
+}
+
+func aggregateReady(objectStatuses []appv1beta1.ObjectStatus) (bool, int) {
+	countReady := 0
+	for _, os := range objectStatuses {
+		if os.Status == StatusReady {
+			countReady++
+		}
+	}
+	if countReady == len(objectStatuses) {
+		return true, countReady
+	}
+	return false, countReady
+}
+
+func (r *ApplicationReconciler) updateApplicationStatus(ctx context.Context, nn types.NamespacedName, status *appv1beta1.ApplicationStatus) error {
+	if err := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		original := &appv1beta1.Application{}
+		if err := r.Get(ctx, nn, original); err != nil {
+			return err
+		}
+		original.Status = *status
+		if err := r.Client.Status().Update(ctx, original); err != nil {
+			return err
+		}
+		return nil
+	}); err != nil {
+		return fmt.Errorf("failed to update status of Application %s/%s: %v", nn.Namespace, nn.Name, err)
+	}
+	return nil
+}
+
+func (r *ApplicationReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	c, err := ctrl.NewControllerManagedBy(mgr).
+		Named("application-controller").
+		For(&appv1beta1.Application{}).Build(r)
```

Comment:
When the application object changed, will do the reconcile to modify the status of the application. But if the deployments (service or strategy and so on)  which belongs to the application changed, should trigger to modify the application. How to guarantee this situation?

Reply:
https://github.com/kubesphere/kubesphere/pull/3403/files#diff-8704122abcfce9d13573e5e86a42be45da623b4e5ff6a5f1ad7176bbf9a5d3a5L73-L97 
Didn't remove code from the last PR

Reply:
Cannot locate the code by your link.

Reply:
Just about a few lines below here

```
for _, s := range sources {
		// Watch for changes to Application
		err = c.Watch(
			&source.Kind{Type: s},
			&handler.EnqueueRequestsFromMapFunc{ToRequests: handler.ToRequestsFunc(
				func(h handler.MapObject) []reconcile.Request {
					return []reconcile.Request{{NamespacedName: types.NamespacedName{
						Name:      servicemesh.GetApplictionName(h.Meta.GetLabels()),
						Namespace: h.Meta.GetNamespace()}}}
				})},
			predicate.Funcs{
				UpdateFunc: func(e event.UpdateEvent) bool {
					return isApp(e.MetaOld, e.MetaNew)
				},
				CreateFunc: func(e event.CreateEvent) bool {
					return isApp(e.Meta)
				},
				DeleteFunc: func(e event.DeleteEvent) bool {
					return isApp(e.Meta)
				},
			})
		if err != nil {
			return err
		}
	}
```